src/Microsoft.DotNet.XHarness.Tests.Runners/xUnit/XUnitFilter.cs

```diff
@@ -71,29 +80,176 @@ public static XUnitFilter CreateClassFilter(string className, bool exclude, stri
             };
         }
 
-        public static XUnitFilter CreateTraitFilter(string traitName, string traitValue, bool exclude)
+        public static XUnitFilter CreateTraitFilter(string traitName, string? traitValue, bool exclude)
         {
-            if (String.IsNullOrEmpty(traitName))
+            if (string.IsNullOrEmpty(traitName))
                 throw new ArgumentException("must not be null or empty", nameof(traitName));
 
             return new XUnitFilter
             {
                 AssemblyName = null,
                 SelectorName = traitName,
-                SelectorValue = traitValue ?? String.Empty,
+                SelectorValue = traitValue ?? string.Empty,
                 FilterType = XUnitFilterType.Trait,
                 Exclude = exclude
             };
         }
 
+        bool ApplyTraitFilter(ITestCase testCase, Func<bool, bool>? reportFilteredTest = null)
+        {
+            Func<bool, bool> log = (result) => reportFilteredTest?.Invoke(result) ?? result;
+
+            if (!testCase.HasTraits())
+            {
+                return log (!Exclude);
+            }
+
+            if (testCase.TryGetTrait(SelectorName!, out var values))
```

Comment:
Should this `!` be handled?

Reply:
Good point, will add that, was added because we do not know if the client will have the nullability enabled.